fish-rust/src/builtins/string/repeat.rs

```diff
@@ -0,0 +1,139 @@
+use super::{Arguments, StringError, StringSubCommand};
+use crate::builtins::shared::{io_streams_t, STATUS_CMD_ERROR, STATUS_CMD_OK};
+use crate::ffi::parser_t;
+use crate::wchar::{wstr, WString, L};
+use crate::wchar_ext::WExt;
+use crate::wgetopt::{wopt, woption, woption_argument_t::*};
+use crate::wutil::fish_wcstol;
+
+#[derive(Default)]
+pub struct Repeat {
+    count: usize,
+    max: usize,
+    quiet: bool,
+    no_newline: bool,
+}
+
+impl StringSubCommand<'_> for Repeat {
+    const LONG_OPTIONS: &'static [woption<'static>] = &[
+        wopt(L!("count"), required_argument, 'n'),
+        wopt(L!("max"), required_argument, 'm'),
+        wopt(L!("quiet"), no_argument, 'q'),
+        wopt(L!("no-newline"), no_argument, 'N'),
+    ];
+    const SHORT_OPTIONS: &'static wstr = L!(":n:m:qN");
+
+    fn parse_opt(&mut self, optarg: Option<&wstr>, c: char) -> Result<(), StringError> {
+        match c {
+            'n' => self.count = fish_wcstol(optarg.unwrap())?.try_into()?,
+            'm' => self.max = fish_wcstol(optarg.unwrap())?.try_into()?,
+            'q' => self.quiet = true,
+            'N' => self.no_newline = true,
+            _ => return Err(StringError::UnknownOption),
+        }
+        return Ok(());
+    }
+
+    fn handle(
+        &mut self,
+        _parser: &mut parser_t,
+        streams: &mut io_streams_t,
+        optind: &mut usize,
+        args: &[&wstr],
+    ) -> Option<libc::c_int> {
+        if self.max == 0 && self.count == 0 {
+            // XXX: This used to be allowed, but returned 1.
+            // Keep it that way for now instead of adding an error.
+            // streams.err.append(L"Count or max must be greater than zero");
+            return STATUS_CMD_ERROR;
+        }
+
+        let mut all_empty = true;
+        let mut first = true;
+        let mut print_newline = true;
+
+        for (w, want_newline) in Arguments::new(args, optind, streams) {
+            print_newline = want_newline;
+            if w.is_empty() {
+                continue;
+            }
+
+            all_empty = false;
+
+            if self.quiet {
+                // Early out if we can - see #7495.
+                return STATUS_CMD_OK;
+            }
+
+            if !first {
+                streams.out.append1('\n');
+            }
+            first = false;
+
+            // The maximum size of the string is either the "max" characters,
+            // or it's the "count" repetitions, whichever ends up lower.
+            let max = if self.max == 0 || (self.count > 0 && w.len() * self.count < self.max) {
```

Comment:
Hmm, overflowing when we have a `max` present and `max <= w.len().checked_mul(self.count).unwrap_or(usize::MAX)` seems like fine behaviour, but if we end up storing an overflowed value should actually be disallowed

I will add a wrapping_mul and comment about how this actually should be, and add
```
# count here is isize::MAX
# we store what to print as usize, so this will overflow
# but we limit it to less than whatever the overflow is
# so this should be fine
string repeat -m1 -n 9223372036854775807 aa
# CHECK: a

# FIXME: handle overflowing nicely
# count is here i64::MAX / 2 + 1
# we end up overflowing, and the result is 0
# but this should work fine, as we limit it way before the overflow
# string repeat -m1 -n 4611686018427387904 aaaa
# DONTCHECKERR: a
```

The second example currently prints nothing

Reply:
Ugh, overflow behaviour is not the same on 32bit since usize, I'm just gonna ignore the tests for now https://github.com/fish-shell/fish-shell/actions/runs/5641883643/job/15280694315?pr=9854